### DISCUSSION TITLE
Blazor is too closed/hard to customize? My experience trying to make a custom Hash Router

### DISCUSSION THREAD

**Discussion by datvm** [dotnet/aspnetcore#49572] [General]
After see #6175, I want to implement my own solution. However, what I thought would be a fun side project turned into frustration. My plan was to simply transform the URL path into hash at the browser end so we can reuse everything else of the current routing system.

- First, I tried extending [`Router`](https://github.com/dotnet/aspnetcore/blob/11083b2154c8e3b7ee2ab6d3e4d1a90b83fb8dbc/src/Components/Components/src/Routing/Router.cs#L17). Nope most of the things are `internal` or `private` and there is not even a way to wrap it due to all the injection. Wrapping: not possible, copying the whole code: not possible due to multiple `internal` classes as well.
- Then I tried a "nuclear" option by replacing Javascript `globalThis.location` (`href`) and `globalThis.history` (for `pushState` and `replaceState`). I failed because `location` cannot be modified in any way: `Proxy`, `defineProperty` or even straight up replacing the `window` object (strangely enough it wasn't a problem with `history`). Then I go to a milder option of replacing [`Blazor._internal.navigationManager`](https://github.com/dotnet/aspnetcore/blob/11083b2154c8e3b7ee2ab6d3e4d1a90b83fb8dbc/src/Components/Web.JS/src/Services/NavigationManager.ts) but still wasn't successful because beside the obvious functions (that barely even called), a few of them can't be modified at all.

So, IMO, for something as important as Routing, it's so hard to intercept and customize. I had a similar frustration before when trying to modify the `IJSRuntime` object (though in the end I barely made it through by modifying the whole DI thing and wrap it). I wish they either open up more point of customization or stop hiding so many important classes.

What do you think? Do you have any suggestion where I may be able to continue with this? Thanks.

**Comment by ahjephson** [comment]
I ran into the same issue recently and ended up solving it by replacing NavigationManager rather than trying to replace Router or intercept everything at the browser level.

That let me keep the real browser path fixed at the app base path and move the Blazor route into the URL fragment instead, so refresh/deep-link scenarios work even when the host can’t rewrite arbitrary paths back to index.html.

I’ve packaged that approach here in case it helps anyone else hitting the same limitation: Blazor.HashRouting on NuGet - https://www.nuget.org/packages/Blazor.HashRouting
